### PR TITLE
Cache all the animator's paths CRC before resolving them

### LINKED ISSUE #853: [Bug]: Stubbed property paths are exported in several animation clips in Bug Fables
### Are you on the latest version of AssetRipper?

Yes, I'm on the latest release of AssetRipper.

### Which release are you using?

Linux x64

### Which game did this occur on?

Bug Fables: The Everlasting Sapling

Steamdb: https://steamdb.info/app/1082710/

### Which Unity version did this occur on?

2018.4.12f1

### Is the game Mono or IL2Cpp?

Mono

### Describe the issue.

When I do an export of the game, all animations clips from the game except those used in a specific controller do not have correct property paths.

Specifically, if I check in the Assets/Ressources/animationcontrollers/bee directory after export, every .anim in there messes with the m_Sprite of the Rotater/Sprite path. This is expected and most clips of most controllers under the animationcontrollers are expected to do this given what I know of the game's infrastructure.

However, if I check any other .anim files under that animationcontrollers folder, none of them features this path, but I am seeing the string "path_904830837" which is notably the same number that appears in ALL the files where this path applies including the good ones.

Here's an example of a bad one:
[100.txt](https://github.com/AssetRipper/AssetRipper/files/11715894/100.txt)

And here's a good one:
[100.txt](https://github.com/AssetRipper/AssetRipper/files/11715895/100.txt)

### Relevant log output

[AssetRipper.log](https://github.com/AssetRipper/AssetRipper/files/11715897/AssetRipper.log)

### PR DESCRIPTION
**Pull request by aldelaro5**
Supersedes parts of #856 

This pr changes the way we approach animation clips's paths resolution. It includes a relatively large refactor that removes TOSBuilder to move its logic into PathChecksumCache (since it seemed underutilised and seemed to do what I wanted it to do better than TOSBuilder which is now less well named).

Previously, we tried to recover the path from the CRC with the assumption that a given clip MUST have an animator or animations with a bound controller containing the clip. While this might be true and it could allow easier recovery, tests and reversing of unity shows that cannot be guaranteed due to the runtime features of animation controllers. What unity does at runtime is whenever the controller is set to the animator and enabled, it will gather all possible path from its GO down the hierarchy to build a CRC cache and use that cache to resolve the bindings to the transforms.

This gives AssetRipper 3 potential scenarios to handle instead of 1:
1. The clip is contained in a controller bound to an animator (what AssetRipper currently assumes)
2. The clip is contained in a controller that has no animator bound to it, but an animator exists at the GO meant to receive the controller at runtime (currently fallsback to `path_`, but recoverable)
3. The clip is not contained in a controller or it is, but no animator exists at the GO meant to receive it (currently fallsback to `path_`, this is irrecoverable since everything happens at runtime)

Scenario 2 is where most of the reported issues about `path_` are and they could be recovered. However, it turns out there is something better than nothing for scenario 3: we can now reverse a CRC string which is most likely incorrect, but technically can work better than nothing.

The way this pr achieves this is by turning around what we do: instead of building the crc cache for each clip and rely on it, this pr builds a global crc cache before processing any clips and from there, every clips uses that cache. This happens to cover scenario 1 implicitly because if it happens, it implies an animator must be present there.

The complexity is O(N) because building the cache is O(N), but since we use a dictionary, looking up a crc is O(1) afterwards. This differs from #856 because that pr was still using the same structure TOSBuilder had which effectively made it per clip per animator or O(N)^2. Finally, if this cannot be found in the cache, we falback to a reversed CRC instead.

Fixes
#853 (another part of it not mentioned will be fixed in another pr)
parts of #954 (the rest is fixed by another pr)

cc: @o5zereth

### CODE REVIEW

**Comment by o5zereth** [comment]
Nice! :D

File: Source/AssetRipper.Processing/AnimationClips/PathChecksumCache.cs [hunk 1]
```diff
@@ -21,10 +27,79 @@ public PathChecksumCache(IAssemblyManager assemblyManager)
...
+	
+	private void AddAvatarTOS(IAvatar avatar)
+	{
+		foreach (uint key in avatar.TOS_C90.Keys)
+		{
+			AddKeys(key, avatar.TOS_C90[key]);
+		}
+	}
+	
+	public void BuildPathsCache(AnimationCache cache)
```
Comment (ds5678) [inline Source/AssetRipper.Processing/AnimationClips/PathChecksumCache.cs:85]: Use `GameBundle` as the parameter and delete the `AnimationCache` class.

**Review by ds5678 — CHANGES_REQUESTED** [review]
Just a start so you can continue working.

File: Source/AssetRipper.Processing/AnimationClips/AnimationClipConverter.cs [hunk 2]
```diff
@@ -551,15 +550,15 @@ private static int GetNextCurve(StreamedFrame frame, int currentCurve)
...
 		{
-			if (tos.TryGetValue(hash, out string? path))
+			if (m_checksumCache.TryGetPath(hash, out string? path))
 			{
 				return path;
 			}
 			else
 			{
-				return UnknownPathPrefix + hash;
+				return Crc32Algorithm.ReverseAscii(hash, UnknownPathPrefix);
```
Comment (ds5678) [inline Source/AssetRipper.Processing/AnimationClips/AnimationClipConverter.cs:561]: Can we do this instead? 
```cs
Crc32Algorithm.ReverseAscii(hash, $"{UnknownPathPrefix}0x{hash:X}_")
```
That would embed the original hash in the output.

**Review by ds5678 — CHANGES_REQUESTED** [review]
Looking great so far.

**Review by ds5678 — APPROVED** [review]

**Pull request merged**